dynamic_config/dynamic_config_manager.py

```diff
@@ -192,19 +193,156 @@ def serialize(self) -> str:
     return self._active_config.SerializeToString()
 
 
+class DynamicEndpointsConfigManager(DynamicConfigManager):
+  """Encapsulates dynamic endpoint configuration."""
+
+  def __init__(self, config: endpoints_config_manager_pb2.DynamicEndpointsConfigManagerSettings):
+    """Create a DynamicEndpointsConfigManager."""
+    DynamicConfigManager.__init__(self)
+    self._active_config = discovery_pb2.DiscoveryResponse()
+    self._active_config.version_info = "1"
+    self._active_config.resources.add()
+
+    # The endpoints to publish
+    self._cluster_load_assignment = endpoint_pb2.ClusterLoadAssignment()
+    self._cluster_load_assignment.cluster_name = config.cluster.name
+    self._cluster_load_assignment.endpoints.add()
+
+    # Convience hook to add endpoints.
+    self._lb_endpoints = self._cluster_load_assignment.endpoints[0].lb_endpoints
+
+    endpoints = self._parseAvailableEndpoints(config.cluster.endpoints)
+    self._inactive_endpoints = []
+    self._randomlyAssignInitialEndpoints(endpoints)
+    # TODO(kbaichoo): refactor into a shared constants file.
+    self._output_file = config.output_file if config.output_file else "new_eds.pb"
+    # TODO(kbaichoo): Handle nanos get ignored
+    self._refresh_interval = config.refresh_interval.ToSeconds()
+    # Tracks the last mutated action. Used for testing.
+    self._last_mutate_action = None
+
+  def _parseAvailableEndpoints(
+      self,
+      endpoints: list[endpoints_config_manager_pb2.DynamicEndpointsConfigManagerSettings.Endpoint]
+  ) -> list[endpoint_components_pb2.LbEndpoint]:
+    """Parse the available endpoints."""
+    parsed_endpoints = []
+    for endpoint_config in endpoints:
+      host = endpoint_components_pb2.LbEndpoint()
+      socket_address = host.endpoint.address.socket_address
+      socket_address.address = endpoint_config.ip
+      socket_address.port_value = int(endpoint_config.port)
+      parsed_endpoints.append(host)
+    return parsed_endpoints
+
+  def _randomlyAssignInitialEndpoints(self, endpoints: list[endpoint_components_pb2.LbEndpoint]):
+    while endpoints:
+      add_as_active_endpoint = random.choice([True, False])
+      if add_as_active_endpoint:
+        self._lb_endpoints.append(endpoints.pop())
+      else:
+        self._inactive_endpoints.append(endpoints.pop())
+
+  def mutate(self):
+    """Invoke to randomly mutate the current endpoint configuration."""
```

Comment:
Since this method can raise an exception, can we document so in the Raises: section of its docstring?

https://google.github.io/styleguide/pyguide.html#24-exceptions

Reply:
done